### DISCUSSION TITLE
What are best practices for logging from lua?

### DISCUSSION THREAD

**Discussion by errael** [neovim/neovim#30770] [Q&A] [answered]
`vim` has `ch_log()`, is there something similar in `neovim`?

At `:help log` it says
```
...Nvim keeps a general log file for internal debugging, plugins and RPC clients...
```
Is the idea to open `$NVIM_LOG_FILE` and append to it directly? No API?

Came across `vim.notify()`. I see
```
This function can be overridden by plugins...
```
I tried `vim.notify = MyNotify` and it works; is that a sledgehammer? Is there a more constrained/appropriate way to handle this? (really need to study `lua` more)

Saw `nvim_echo` in a test in a PR; looks interesting. Wonder if some things for `{opts}`, like
- `log` - that means send output to `neovim`'s log
- `callback` - dispatch to this function

would be useful. These could have the flavor of unix's tee.

**Comment by justinmk** [comment]
> `vim` has `ch_log()`, is there something similar

Planning to revisit that next year. https://github.com/neovim/neovim/pull/7062

`nvim_echo` and `vim.notify` are both half-baked and need to be revisited.

**Comment by errael** [comment]
There's also options `'debug'` and `'verbose'`. I wonder if/how these could play into logging.

**Comment by errael** [comment]
A common use case for me is doing `tail -f log_file` while debugging/testing. Unbuffered output is important.

After some playing around, starting up with

```
nvim -Vlog_file ...
```

and doing

```lua
vim.api.nvim_echo({{"message\n"}}, false, {verbose = true})
```

is very handy. While most of the stuff in `log_file` goes out buffered. The text from `nvim_echo` is **unbuffered**, (hooray) and works well with `tail -f log_file`.

**Comment by errael** [comment, marked as answer]
Closing this. Monitoring #35558.
